File: 08-tries/00211-Design_Add_And_Search_Words_Data_Structure/00211-design_add_and_search_words_data_structure.go
// 211: Design Add And Search Words Data Structure
// https://leetcode.com/problems/design-add-and-search-words-data-structure/

package main

import "fmt"


type WordDictionary struct {
    children [26]*WordDictionary
	isEndOfWord bool
}


func Constructor() WordDictionary {
	return WordDictionary{}
}


func (this *WordDictionary) AddWord(word string)  {
    current := this

	for _, c := range word {
		if current.children[c - 'a'] == nil {
			current.children[c - 'a'] =  &WordDictionary{}
		}
		current = current.children[c - 'a'];
	}

	current.isEndOfWord = true;
}


func (this *WordDictionary) Search(word string) bool {
	current := this

	for i:=0; i<len(word); i++ {
		c := word[i]

		if c == '.' {
			for _, ch := range current.children {
				if ch!=nil && ch.Search(word[i+1:]) {return true}
			}
			return false
		}

		if (current.children[c - 'a'] == nil) {return false}
		current = current.children[c - 'a'];
	}

	return current!=nil && current.isEndOfWord
}


func main() {
	wd := Constructor()

	// OPERATIONS
	wd.AddWord("bad")
    wd.AddWord("dad")
    wd.AddWord("mad")
    fmt.Println(wd.Search("pad"))
    fmt.Println(wd.Search("bad"))
    fmt.Println(wd.Search(".ad"))
    fmt.Println(wd.Search("b.."))
}

File: 08-tries/00211-Design_Add_And_Search_Words_Data_Structure/00211-design_add_and_search_words_data_structure.py
# 211: Design Add And Search Words Data Structure
# https://leetcode.com/problems/design-add-and-search-words-data-structure/



class WordDictionary:
    def __init__(self):
        self.children = [None]*26
        self.isEndOfWord = False

    def addWord(self, word: str) -> None:
        current = self

        for c in word:
            if current.children[ord(c) - ord('a')] == None:
                current.children[ord(c) - ord('a')] = WordDictionary()
            current = current.children[ord(c) - ord('a')]
        
        current.isEndOfWord = True
        

    def search(self, word: str) -> bool:
        current = self

        for i in range(len(word)):
            c = word[i]
            if c == '.':
                for ch in current.children:
                    if ch != None and ch.search(word[i+1:]): return True
                return False
            
            if current.children[ord(c) - ord('a')] == None: return False
            current = current.children[ord(c) - ord('a')]
        
        return current != None and current.isEndOfWord
    


if __name__ == "__main__":
    wd = WordDictionary()

    # OPERATIONS
    wd.addWord("bad");
    wd.addWord("dad");
    wd.addWord("mad");
    print(wd.search("pad"))
    print(wd.search("bad"))
    print(wd.search(".ad"))
    print(wd.search("b.."))
